pkg/crdschema/gen.go

```diff
@@ -0,0 +1,377 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	crdgen "sigs.k8s.io/controller-tools/pkg/crd"
+	crdmarkers "sigs.k8s.io/controller-tools/pkg/crd/markers"
+	crdschemayaml "sigs.k8s.io/controller-tools/pkg/crdschema/yaml"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func init() {
+	utilruntime.Must(apiextensionsv1beta1.AddToScheme(scheme))
+}
+
+// Generator is a genall.Generator that generates CRD schemas.
+type Generator struct {
+	// ManifestsPath is a path with CustomResourceDefinition manifest files
+	ManifestsPath string `marker:"manifests,optional"`
+	// If true, no file is written. The Generate func will return an
+	// error if a file on disk differs from the generated manifest.
+	VerifyOnly bool `marker:",optional"`
+}
+
+var _ genall.Generator = &Generator{}
+
+func (Generator) RegisterMarkers(into *markers.Registry) error {
+	return crdmarkers.Register(into)
+}
+
+func (g Generator) Generate(ctx *genall.GenerationContext) (result error) {
+	p := &crdgen.Parser{
+		Collector: ctx.Collector,
+		Checker:   ctx.Checker,
+	}
+	f := crdgen.Flattener{
+		Parser: p,
+	}
+
+	crdgen.AddKnownTypes(p)
+	for _, root := range ctx.Roots {
+		p.NeedPackage(root)
+	}
+
+	metav1Pkg := crdgen.FindMetav1(ctx.Roots)
+	if metav1Pkg == nil {
+		// no objects in the roots, since nothing imported metav1
+		fmt.Printf("No metav1 import found in any of the API packages.\n")
+		return nil
+	}
+
+	// load existing manifests from manifests/ dir
+	existingCRDsByFileNames, err := crdsFromDirectory(ctx, g.ManifestsPath)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Found %d CRD manifests.\n", len(existingCRDsByFileNames))
+	existingFileNamesByGroupKind := map[schema.GroupKind]string{}
+	for fn, crd := range existingCRDsByFileNames {
+		existingFileNamesByGroupKind[crd.GroupKind] = fn
+	}
+
+	changedFiles := []string{}
+	for _, groupKind := range crdgen.FindKubeKinds(p, metav1Pkg) {
+		fn, found := existingFileNamesByGroupKind[groupKind]
+		if !found {
+			continue
+		}
+		existing := existingCRDsByFileNames[fn]
+
+		// get existing versions
+		existingVersions := make(map[string]bool, 0)
+		if existing.CRD.Spec.Version != "" {
+			existingVersions[existing.CRD.Spec.Version] = true
+		}
+		for _, v := range existing.CRD.Spec.Versions {
+			existingVersions[v.Name] = true
+		}
+
+		// go through source code versions
+		versionSchemas := map[string]*apiextensionsv1beta1.JSONSchemaProps{}
+		for pkg, gv := range p.GroupVersions {
+			if gv.Group != groupKind.Group {
+				continue
+			}
+			if !existingVersions[gv.Version] {
+				fmt.Printf("Source code for %s in %s/%s not found.\n", groupKind.Kind, gv.Group, gv.Version)
```

Comment:
Don't `Printf` to log errors, and especially don't `Printf` to stdout.

Reply:
done